File: src/test/java/mediathek/daten/IndexedFilmListTest.java
package mediathek.daten;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.List;

class IndexedFilmListTest {

    @Test
    void getFilmByFilmNr_returnsMatchingFilmAfterIndexBuild() {
        var list = new IndexedFilmList();
        var firstFilm = createFilm("Alpha");
        var secondFilm = createFilm("Beta");

        list.addAll(List.of(firstFilm, secondFilm));

        Assertions.assertSame(firstFilm, list.getFilmByFilmNr(firstFilm.getFilmNr()));
        Assertions.assertSame(secondFilm, list.getFilmByFilmNr(secondFilm.getFilmNr()));
    }

    @Test
    void getFilmByFilmNr_rebuildsIndexAfterAdd() {
        var list = new IndexedFilmList();
        var firstFilm = createFilm("Alpha");
        list.add(firstFilm);

        Assertions.assertSame(firstFilm, list.getFilmByFilmNr(firstFilm.getFilmNr()));

        var secondFilm = createFilm("Beta");
        list.add(secondFilm);

        Assertions.assertSame(secondFilm, list.getFilmByFilmNr(secondFilm.getFilmNr()));
    }

    @Test
    void getFilmByFilmNr_rebuildsIndexAfterRemoveAndClear() {
        var list = new IndexedFilmList();
        var firstFilm = createFilm("Alpha");
        var secondFilm = createFilm("Beta");
        list.addAll(List.of(firstFilm, secondFilm));

        Assertions.assertSame(firstFilm, list.getFilmByFilmNr(firstFilm.getFilmNr()));

        list.remove(firstFilm);

        Assertions.assertNull(list.getFilmByFilmNr(firstFilm.getFilmNr()));
        Assertions.assertSame(secondFilm, list.getFilmByFilmNr(secondFilm.getFilmNr()));

        list.clear();

        Assertions.assertNull(list.getFilmByFilmNr(secondFilm.getFilmNr()));
    }

    @Test
    void getFilmByFilmNr_rebuildsIndexAfterSet() {
        var list = new IndexedFilmList();
        var originalFilm = createFilm("Original");
        list.add(originalFilm);

        Assertions.assertSame(originalFilm, list.getFilmByFilmNr(originalFilm.getFilmNr()));

        var replacementFilm = createFilm("Replacement");
        list.set(0, replacementFilm);

        Assertions.assertNull(list.getFilmByFilmNr(originalFilm.getFilmNr()));
        Assertions.assertSame(replacementFilm, list.getFilmByFilmNr(replacementFilm.getFilmNr()));
    }

    private DatenFilm createFilm(String title) {
        var film = new DatenFilm();
        film.setSender("ARD");
        film.setThema("Test");
        film.setTitle(title);
        film.setNormalQualityUrl("https://example.invalid/" + title);
        return film;
    }
}

File: src/test/kotlin/mediathek/daten/IndexedFilmListTest.kt
package mediathek.daten

import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Test

internal class IndexedFilmListTest {

    @Test
    fun `getFilmByFilmNr returnsMatchingFilmAfterIndexBuild`() {
        val list = IndexedFilmList()
        val firstFilm = createFilm("Alpha")
        val secondFilm = createFilm("Beta")

        list.addAll(listOf(firstFilm, secondFilm))

        assertSame(firstFilm, list.getFilmByFilmNr(firstFilm.filmNr))
        assertSame(secondFilm, list.getFilmByFilmNr(secondFilm.filmNr))
    }

    @Test
    fun `getFilmByFilmNr rebuildsIndexAfterAdd`() {
        val list = IndexedFilmList()
        val firstFilm = createFilm("Alpha")
        list.add(firstFilm)

        assertSame(firstFilm, list.getFilmByFilmNr(firstFilm.filmNr))

        val secondFilm = createFilm("Beta")
        list.add(secondFilm)

        assertSame(secondFilm, list.getFilmByFilmNr(secondFilm.filmNr))
    }

    @Test
    fun `getFilmByFilmNr rebuildsIndexAfterRemoveAndClear`() {
        val list = IndexedFilmList()
        val firstFilm = createFilm("Alpha")
        val secondFilm = createFilm("Beta")
        list.addAll(listOf(firstFilm, secondFilm))

        assertSame(firstFilm, list.getFilmByFilmNr(firstFilm.filmNr))

        list.remove(firstFilm)

        assertNull(list.getFilmByFilmNr(firstFilm.filmNr))
        assertSame(secondFilm, list.getFilmByFilmNr(secondFilm.filmNr))

        list.clear()

        assertNull(list.getFilmByFilmNr(secondFilm.filmNr))
    }

    @Test
    fun `getFilmByFilmNr rebuildsIndexAfterSet`() {
        val list = IndexedFilmList()
        val originalFilm = createFilm("Original")
        list.add(originalFilm)

        assertSame(originalFilm, list.getFilmByFilmNr(originalFilm.filmNr))

        val replacementFilm = createFilm("Replacement")
        list[0] = replacementFilm

        assertNull(list.getFilmByFilmNr(originalFilm.filmNr))
        assertSame(replacementFilm, list.getFilmByFilmNr(replacementFilm.filmNr))
    }

    private fun createFilm(title: String): DatenFilm =
        DatenFilm().apply {
            sender = "ARD"
            thema = "Test"
            this.title = title
            setNormalQualityUrl("https://example.invalid/$title")
        }
}
